Source/WebKit/Platform/IPC/HandleMessage.h

```diff
@@ -361,23 +367,38 @@ struct MethodSignatureValidationImpl<std::tuple<MessageArgumentTypes...>, std::t
     static constexpr bool expectsConnectionArgument = false;
     using MessageArguments = std::tuple<std::remove_cvref_t<MessageArgumentTypes>...>;
     using CompletionHandlerType = void;
-    static constexpr bool completionHandlerIsWrapped = false;
+    static constexpr std::identity wrapCompletionHandler { };
+    static constexpr std::identity getCompletionHandlerArgument { };
 };
 
+// Properties, traits and conversions for C++ message handler functions
 template<typename... MessageArgumentTypes, typename... CompletionHandlerArgumentTypes>
 struct MethodSignatureValidationImpl<std::tuple<MessageArgumentTypes...>, std::tuple<CompletionHandler<void(CompletionHandlerArgumentTypes...)>&&>>
     : MethodSignatureValidationImpl<std::tuple<MessageArgumentTypes...>, std::tuple<>> {
     using CompletionHandlerArguments = std::tuple<std::remove_cvref_t<CompletionHandlerArgumentTypes>...>;
     using CompletionHandlerType = CompletionHandler<void(CompletionHandlerArgumentTypes...)>;
-    static constexpr bool completionHandlerIsWrapped = false;
+    static constexpr std::identity wrapCompletionHandler { };
+    static constexpr std::identity getCompletionHandlerArgument { };
 };
 
+// Properties, traits and conversions for Swift message handler functions
+// (exposed over Swift/C++ interop)
 template<typename... MessageArgumentTypes, typename... CompletionHandlerArgumentTypes>
 struct MethodSignatureValidationImpl<std::tuple<MessageArgumentTypes...>, std::tuple<WTF::RefCountable<WTF::CompletionHandler<void(CompletionHandlerArgumentTypes...)>>*>>
     : MethodSignatureValidationImpl<std::tuple<MessageArgumentTypes...>, std::tuple<>> {
     using CompletionHandlerArguments = std::tuple<std::remove_cvref_t<CompletionHandlerArgumentTypes>...>;
-    using CompletionHandlerType = WTF::RefCountable<WTF::CompletionHandler<void(CompletionHandlerArgumentTypes...)>>;
-    static constexpr bool completionHandlerIsWrapped = true;
+    using CompletionHandlerType = CompletionHandler<void(CompletionHandlerArgumentTypes...)>;
+    template<typename CH>
+    static Ref<WTF::RefCountable<WTF::CompletionHandler<CH>>> wrapCompletionHandler(CompletionHandler<CH>&& handler)
```

Comment:
Not new in this patch, but it's curious that we need to wrap in `RefCountable` before calling into Swift. Swift supports move-only types via `borrowing` and `consuming`. Did we try that?

Reply:
Not yet supported by reverse interop, sadly - rdar://162361370